### ISSUE TITLE
Change project structure to decouple querqy-lucene from querqy-solr releases

### ISSUE THREAD

**Issue by renekrie** [querqy/querqy#304]
With the split of the Apache Lucene and Apache Solr projects, we can no longer rely on a Solr version becoming available soon after a Lucene release. Currently, this is blocking Querqy for Elasticsearch (https://github.com/querqy/querqy-elasticsearch). We cannot create a Querqy version for Elasticsearch at the moment as ES 8 depends on Lucene 9 and as our releases of the querqy-lucene package is tied to releases of querqy-solr, which is dependent on Solr and its releases. As we don't know, when Solr 9 will be released we should decouple querqy-lucene from querqy-solr releases.

As a first approach, we could simply remove the querqy-for-lucene parent (which has querqy-lucene and querqy-solr as modules) and make querqy-solr dependent on the released version of querqy-lucene (Lucene 8).

**Comment by ps48** [comment]
**Update:**[ Solr 9](https://solr.apache.org/downloads.html) was recently released. I think we should start the decoupling the project and look into version upgrade.

**Comment by renekrie** [comment]
Just to say that I have started work on the Lucene/Solr 9 version and I think I can share this tomorrow. Basically there aren't too many changes in Lucene/Solr and the querqy-lucene module already passes tests. querqy-solr compiles but there is an issue with getting the Solr testing framework running. We can then decide whether we want to decouple before the release or thereafter but I wanted to have all tests passing before going ahead -> it might be quicker to first release and then split as the project will have to also accommodate past Solr versions.

**Comment by ps48** [comment]
@renekrie That's amazing! I tried to upgrade [querqy-lucene](https://github.com/ps48/querqy/tree/lucene_9) on my side. I saw some PRMS test failures. With respect to "split and upgrade Vs. upgrade and split" we can just check what's most easy thing for the current users(other libraries and plugins) of querqy. From a "plugin using querqy" perspective, it would make sense to upgrade first as changing to Lucene 9 might already have some additional overhead for the plugin. I am not aware of Solr side of things.

**Comment by renekrie** [comment]
> @renekrie That's amazing! I tried to upgrade [querqy-lucene](https://github.com/ps48/querqy/tree/lucene_9) on my side. I saw some PRMS test failures. 

Yeah, DisMaxQuery clauses are no longer guaranteed to be in a stable order - which is an impediment for our tests but the underlying implementation of PRMS is still okay.

**Comment by renekrie** [comment]
@ps48 One thing where we could probably need helping hands for the split: I'm not sure what our test coverage will look like for querqy-lucene once we make it more independent of querqy-solr. We have a number of tests via the Solr testing framework that work like small 'use case tests' and obviously they also cover some querqy-lucene code. I think overall the coverage in querqy-lucene will still look okay but we should check this and maybe add a few tests where needed.

**Comment by JohannesDaniel** [comment]
One additional question: The idea of this is to release querqy-lucene fully independent of any search engine integration?

**Comment by renekrie** [comment]
> One additional question: The idea of this is to release querqy-lucene fully independent of any search engine integration?

Yes. In a first step, we could still keep everything in this repo - similar to what we are doing with querqy-core.

**Comment by renekrie** [comment]
> I think test coverage indeed might be a problem as these tests have always been some kind of "full picture tests" for complex issues and interactions across components. Was just wondering whether we should brainstorm about some kind of abstraction in those tests that make the tests search engine agnostic? I think the majority of these tests have been implemented in very similar ways using SolrQueryRequest and assertq.

We do have some support using the matchers in `querqy.lucene.rewrite.AbstractLuceneQueryTest`. We might want to add a few utilities for accessing explain output from Lucene Weights. Also, when we work with actual indexes, the randomness that comes with the Lucene testing framework is not always what we want.

**Comment by renekrie** [comment]
This has been implemented. Querqy Solr has its own repo now at https://github.com/querqy/querqy-solr.
